src/common/ina.rs

```diff
@@ -0,0 +1,246 @@
+    status::Status,
+};
+
+const DEFAULT_I2C_ADDRESS: u8 = 0x40;
+const DEFAULT_SHUNT_RESISTANCE_OHMS: f64 = 0.1;
+
+const INA_219_CALIBRATION_SCALE: f64 = 0.04096;
+const INA_226_CALIBRATION_SCALE: f64 = 0.00512;
+
+const DEFAULT_CONFIG_REGISTER_VALUE: u16 = 0x399F;
+const CONFIG_REGISTER: u8 = 0x00;
+const VOLTAGE_REGISTER: [u8; 1] = [0x02];
+const CURRENT_AMPERES_REGISTER: [u8; 1] = [0x04];
+const POWER_REGISTER: [u8; 1] = [0x03];
+
+pub(crate) fn register_models(registry: &mut ComponentRegistry) {
+    if registry
+        .register_power_sensor("ina219", &ina219_from_config)
+        .is_err()
+    {
+        log::error!("gpio model is already registered")
+    }
+    if registry
+        .register_power_sensor("ina226", &ina226_from_config)
+        .is_err()
+    {
+        log::error!("gpio model is already registered")
+    }
+}
+
+pub(crate) fn ina219_from_config(
+    cfg: ConfigType,
+    deps: Vec<Dependency>,
+) -> anyhow::Result<PowerSensorType> {
+    Ok(Arc::new(Mutex::new(from_config(Model::Ina219, cfg, deps)?)))
+}
+
+pub(crate) fn ina226_from_config(
+    cfg: ConfigType,
+    deps: Vec<Dependency>,
+) -> anyhow::Result<PowerSensorType> {
+    Ok(Arc::new(Mutex::new(from_config(Model::Ina226, cfg, deps)?)))
+}
+
+fn from_config(
+    model: Model,
+    cfg: ConfigType,
+    dependencies: Vec<Dependency>,
+) -> anyhow::Result<Ina<I2cHandleType>> {
+    let i2c_address = cfg
+        .get_attribute::<u8>("i2c_address")
+        .unwrap_or(DEFAULT_I2C_ADDRESS);
+
+    let default_max_current_amperes = match &model {
+        Model::Ina219 => 3.2,
+        Model::Ina226 => 20.0,
+    };
+    let max_current_amperes = cfg
+        .get_attribute::<f64>("max_current_amps")
+        .unwrap_or(default_max_current_amperes);
+    let max_current_nano_amperes = (max_current_amperes * 1e9) as i64;
+
+    let shunt_resistance_ohms = cfg
+        .get_attribute::<f64>("shunt_resistance")
+        .unwrap_or(DEFAULT_SHUNT_RESISTANCE_OHMS);
+    let shunt_resistance_nano_ohms = (shunt_resistance_ohms * 1e9) as i64;
+
+    let i2c_name = cfg
+        .get_attribute::<String>("i2c_bus")
+        .map_err(|_| anyhow::anyhow!("i2c_bus is a required config attribute for {}", model))?;
+    let board = get_board_from_dependencies(dependencies).ok_or(anyhow::anyhow!(
+        "a board must be configured to configure an {} sensor",
+        model
+    ))?;
+    let i2c_handle = board.get_i2c_by_name(i2c_name)?;
+
+    Ina::new(
+        model,
+        i2c_handle,
+        i2c_address,
+        max_current_nano_amperes,
+        shunt_resistance_nano_ohms,
+    )
+}
+
+#[derive(Clone, Copy)]
+pub enum Model {
+    Ina219,
+    Ina226,
+}
+
+impl fmt::Display for Model {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        let model_str = match self {
+            Model::Ina219 => "INA219",
+            Model::Ina226 => "INA226",
+        };
+        write!(f, "{model_str}")
+    }
+}
+
+pub struct Ina<H> {
```

Comment:
- Why is this generic over H if the only H that is impl'd is for `I2CHandle`?
- Does it need to be `pub`?